### ISSUE TITLE
--icc-force-contrast does not work with --vo=gpu-next

### ISSUE THREAD

**Issue by aufkrawall** [mpv-player/mpv#9752]
```
icc-profile-auto
icc-force-contrast=inf
```

`--vo=gpu`:
![Screenshot_20220123_202114](https://user-images.githubusercontent.com/13510490/150694527-0606c6ac-0bf8-43ce-84b1-b930059d3a40.png)

`--icc-force-contrast=inf` should turn BT.1886 gamma curve into pure power 2.4, but it doesn't with `--vo=gpu-next`:
![Screenshot_20220123_202028](https://user-images.githubusercontent.com/13510490/150694664-7dad25b7-8183-4740-8822-92e8cdfdcf01.png)

`--vf=format:gamma=gamma2.4` can be used as a workaround.

Log:
[output.txt](https://github.com/mpv-player/mpv/files/7921227/output.txt)

latest mpv & libplacebo git-master as of today.

I btw. still think that pure power gamma 2.4 would be a better default for rec709 content, as BT.1886 just looks too bright with most of the content on any display that doesn't have an extremely low black point.

**Comment by ValZapod** [comment]
vf format also does not work. Tested on webkit display-p3 image. You know the one.

**Comment by ValZapod** [comment]
Also target pq does not work on dolby vision content. Not nice, @haasn?

**Comment by CrHasher** [comment]
Please remove label "os:linux", it applies also to windows. In addition "gamma-factor" doesn't work.

**Comment by OrangeColaJuice** [comment]
Is the "bt.1886" treated like something in-between sRGB and gamma2.2 because it has hard-coded 1000 contrast over the spec's formula? That's definitely how it feels to me after experimenting.

**Comment by haasn** [comment]
> Is the "bt.1886" treated like something in-between sRGB and gamma2.2 because it has hard-coded 1000 contrast over the spec's formula? That's definitely how it feels to me after experimenting.

Yes and no. libplacebo assumes 1000:1 contrast by default for SDR displays with unknown contrast information. If you use an ICC profile, it will instead take the effective contrast of the ICC profile's LUTs. (Note that `--icc-force-contrast` is currently ignored by `--vo=gpu-next`, this is a known limitation.)

In theory BT.1886 can be tuned to any contrast point you want, but the mpv integration does not expose all options.

**Comment by HybridDog** [comment]
If I understand it correctly, the reason why the bt.1886 EOTF depends on the display's minimum and maximum brightness is because dark areas should have a similar contrast on displays with unperfect (e.g. LCD) and perfect (e.g. OLED) black such that on LCDs it is possible to see objects in a dark scene as easily as on a CRT.

Perhaps, instead of applying a display-dependent bt.1886 EOTF, it is possible to use a pure 2.4 gamma and a shader which sharpens edges in dark areas, where the sharpening strength depends on the contrast difference between a display with perfect black and the currently used display. I assume sharpening is a way to make objects easier to distinguish in dark areas.

In comparison to bt.1886, this may make perceptual color hues more consistent across displays.
On the other hand, if the author of the video has created it such that with a bt.1886 EOTF it looks good on various displays, bt.1886 should be the better choice. I guess most videos are from authors who don't bother to do this.

I don't know if my idea of using sharpening instead of display-dependent bt.1886 is good or bad.

**Comment by CrHasher** [comment]
> If I understand it correctly, the reason why the bt.1886 EOTF depends on the display's minimum and maximum brightness is because dark areas should have a similar contrast on displays with unperfect (e.g. LCD) and perfect (e.g. OLED) black such that on LCDs it is possible to see objects in a dark scene as easily as on a CRT.
> 
> Perhaps, instead of applying a display-dependent bt.1886 EOTF, it is possible to use a pure 2.4 gamma and a shader which sharpens edges in dark areas, where the sharpening strength depends on the contrast difference between a display with perfect black and the currently used display. I assume sharpening is a way to make objects easier to distinguish in dark areas.
> 
> In comparison to bt.1886, this may make perceptual color hues more consistent across displays. On the other hand, if the author of the video has created it such that with a bt.1886 EOTF it looks good on various displays, bt.1886 should be the better choice. I guess most videos are from authors who don't bother to do this.
> 
> I don't know if my idea of using sharpening instead of display-dependent bt.1886 is good or bad (probably it's bad).

Not really sure what you mean by sharpening dark areas, you would like to do pure 2.4 and then what exactly? What can a shader do? If you start processing that image after you do pure 2.4 you end up applying another curve most probably, andvtgat makes no sense. Things are not linear so you can't do things in a shader other than making things worse most probably.

**Comment by HybridDog** [comment]
I think after applying the 2.4 gamma, the values are linear.
The sharpening shader would modify the linear values using neighbouring pixels without changing the average brightness at smooth regions.
The linear values would then be mapped to values for the display. Since the display cannot (or not necessarily) display the darkest values, a toe (like in common tonemapping curves) would squash them together near the darkest black supported by the display. Clipping dark values instead of using a toe should also work, but can lead to information loss (in addition to the loss caused by quantisation).

I made an image with the GIMP to illustrate my idea:
![scharfbeispiel](https://github.com/user-attachments/assets/08fd1e90-abcf-4bbb-8d23-886dbb0f505d)

Top: Image with perfect black, i.e. the ideal image

Middle: Image where values are clamped to an elevated black level, i.e. the image where values are the same as the ideal image if they are sufficiently bright. Details are no longer visible in dark areas because of the unperfect darkness.

Bottom: Image where dark areas are sharpened and then values are clamped to an elevated black level, i.e. my suggested idea. The details are visible even though there is unperfect darkness like in the middle image.

**Comment by CrHasher** [comment]
This is what you might end up with IMO:
https://github.com/user-attachments/assets/66fc817c-9314-4926-92c9-cd625b29a5f0
